cli/command/cli.go

```diff
@@ -255,18 +255,26 @@ func NewAPIClientFromFlags(opts *cliflags.CommonOptions, configFile *configfile.
 	if err != nil {
 		return nil, errors.Wrap(err, "unable to resolve docker endpoint")
 	}
-	return newAPIClientFromEndpoint(endpoint, configFile)
+	return newAPIClientFromEndpoint(opts, endpoint, configFile)
 }
 
-func newAPIClientFromEndpoint(ep docker.Endpoint, configFile *configfile.ConfigFile) (client.APIClient, error) {
+func newAPIClientFromEndpoint(opts *cliflags.CommonOptions, ep docker.Endpoint, configFile *configfile.ConfigFile) (client.APIClient, error) {
 	clientOpts, err := ep.ClientOpts()
 	if err != nil {
 		return nil, err
 	}
-	customHeaders := make(map[string]string, len(configFile.HTTPHeaders))
+	customHeaders := make(map[string]string, len(configFile.HTTPHeaders)+len(opts.HttpHeaders))
 	for k, v := range configFile.HTTPHeaders {
 		customHeaders[k] = v
 	}
+
+	for i := 0; i < len(opts.HttpHeaders); i++ {
+		split := strings.Split(opts.HttpHeaders[i], ":")
```

Comment:
I have yet to have a closer look, but some quick thoughts;

- we should probably use `=` as separator (`--http-header My-Header=myvalue`)
- use `strings.SplitN(.., ..., 2)` (as we expect only a single separator); this prevents issues if the value would contain `:`, which could be if (just thinking out lout) for example, an IPv6 address is passed (`Foo-Header=[::]:123`)
- We probably need to do this in some other places as well (), but perhaps we should also use [`http. CanonicalHeaderKey()`](https://pkg.go.dev/net/http#CanonicalHeaderKey) (after stripping whitespace), to prevent duplicate headers, and to prevent `User-Agent` from being overridden (which we don't want); it looks like the [client already does normalising](https://github.com/moby/moby/blob/f1dd6bf84e28930e1ccd903361f9284fb22d3b8a/client/request.go#L245) (I just notice it doesn't trim whitespace, and probably should though), but it doesn't hurt to prevent possible duplicates here

Overall, I still want to look if this function (`newAPIClientFromEndpoint()`) is the best place to set this; I'm considering if perhaps we should merge the config-file and flag options earlier, perhaps as part of [`DockerCLI.Initialize()`](https://github.com/docker/cli/blob/d0df532a25ad18e8b4325edcf46f3d0f05b47708/cli/command/cli.go#L195) ? (I'll have to have a look if that would work)

Reply:
> we should probably use = as separator (--http-header My-Header=myvalue)

I'm trying to keep consistent with [`curl`](https://stackoverflow.com/questions/356705/how-to-send-a-header-using-a-http-request-through-a-curl-call) since it has been widely accepted. And I'm not sure if there are traps if we use `=`

Reply:
> I'm considering if perhaps we should merge the config-file and flag options earlier, perhaps as part of [DockerCLI.Initialize()](https://github.com/docker/cli/blob/d0df532a25ad18e8b4325edcf46f3d0f05b47708/cli/command/cli.go#L195)

Agree. This will make the whole configing procedure more clear

Reply:
> use strings.SplitN(.., ..., 2)

Addressed